doc/sphinx/proof-engine/ltac.rst

```diff
@@ -469,45 +500,117 @@ We can check if a tactic made progress with:
    .. exn:: Failed to progress.
       :undocumented:
 
-Backtracking branching: +
-~~~~~~~~~~~~~~~~~~~~~~~~~
-
-We can branch with the following structure:
+Branching and backtracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. tacn:: @ltac_expr__1 + @ltac_expr__2
+|Ltac| provides several :gdef:`branching` tactics that permit trying multiple alternative tactics
+for a proof step.  For example, :tacn:`first`, which tries several alternatives and selects the first
+that succeeds, or :tacn:`tryif`, which tests whether a given tactic would succeed or fail if it was
+applied and then, depending on the result, applies one of two alternative tactics.  There are
+are also looping constructs :tacn:`do` and :tacn:`repeat`.  The order in which the subparts
+of these tactics are evaluated is generally similar to
+structured programming constructs in many languages.
+
+The :tacn:`+<+ (backtracking branching)>`, :tacn:`multimatch` and :tacn:`multimatch goal` tactics
+provide more complex capability.  Rather than applying a single successful
+tactic, these tactics generate a series of successful tactic alternatives that are tried sequentially
+when subsequent tactics outside these constructs fail.  Please examine the :tacn:`multimatch` example
+:ref:`here<match_vs_multimatch_ex>`.
+
+These constructs are evaluated using :gdef:`backtracking`.  Each  creates a
+:gdef:`backtracking point`.  When a subsequent tactic fails, evaluation continues from the nearest
+prior backtracking point with the next successful alternative and repeats the tactics after
+the backtracking point.  When a backtracking point has
+no more successful alternatives, evaluation continues from the next prior backtracking point.
+If there are no more prior backtracking points, the overall tactic fails.
+
+Thus, backtracking tactics can have multiple successes.  Non-backtracking constructs that appear
+after a backtracking point are reprocessed after backtracking, as in the example
+:ref:`here<match_vs_multimatch_ex>`, in which
+the :tacn:`;<ltac-seq>` construct is reprocessed after backtracking.  When a backtracking construct is within
+a non-backtracking construct, the latter uses the :gdef:`first success`.  Backtracking to
+a point within a non-backtracking construct won't change the branch that was selected by the
+non-backtracking construct.
+
+The :tacn:`once` tactic stops further backtracking to backtracking points within that tactic.
+
+eval vs execution (also out of order)
+
+Branching with backtracking: +
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can branch with backtracking with the following structure:
+
+.. tacn:: @ltac_expr1 + {| @ltac_expr2 | @binder_tactic }
```

Comment:
Why are the `ltac_expr`s here not subscripted?

Reply:
Because that's the definition of the production.  Since the names different, subscripts are not needed.  (Subscripts are more work to maintain when the grammar changes, too.)

```
ltac_expr2: [
| ltac_expr1 "+" [ ltac_expr2 | binder_tactic ]
```